dependencies/mqtt/mqtt.go

```diff
@@ -146,3 +155,87 @@ type ErrorDialer struct{}
 func (d ErrorDialer) Dial(ctx context.Context, brokers []string, options Options) (Client, error) {
 	return nil, errors.New(codes.Invalid, "Dialer.Dial called on an error dependency")
 }
+
+type poolClientOptions struct {
+	brokers string
+	options Options
+}
+
+type poolDialer struct {
+	dialer  Dialer
+	clients map[poolClientOptions]*[]Client
+	mu      sync.Mutex
+}
+
+func newPoolDialer(d Dialer) *poolDialer {
+	return &poolDialer{
+		dialer:  d,
+		clients: make(map[poolClientOptions]*[]Client),
+	}
+}
+
+func (p *poolDialer) Dial(ctx context.Context, brokers []string, options Options) (Client, error) {
+	opts := poolClientOptions{
+		brokers: strings.Join(brokers, ";"),
+		options: options,
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	var client Client
+	if clients := p.clients[opts]; clients != nil && len(*clients) > 0 {
+		client = (*clients)[len(*clients)-1]
+		*clients = (*clients)[:len(*clients)-1]
+	} else {
+		var err error
+		client, err = p.dialer.Dial(ctx, brokers, options)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &poolClient{
+		Client:  client,
+		options: opts,
+		pool:    p,
+	}, nil
+}
+
+func (p *poolDialer) Close() error {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	var firstErr error
+	for _, clients := range p.clients {
+		for _, client := range *clients {
+			if err := client.Close(); err != nil && firstErr == nil {
+				firstErr = err
```

Comment:
If we only show the first error, are we potentially hiding other errors on `client.Close`?